single_offer/src/lib.rs

```diff
@@ -13,202 +12,146 @@ mod token {
-    // the seller and the buyer need to transfer the sell_token and buy_token to this contract prior to calling
-    // trade. Due to this and the fact that the buyer is a parameter to trade, the buyer must tranfer the buy_token
-    // to the contract and call trade in the same transaction for safety.
-    fn trade(e: Env, to: Identifier, min: i128);
-
-    // Sends amount of sell_token from this contract to the admin. Must be authorized by admin
-    fn withdraw(e: Env, amount: i128);
-
-    // Updates the price. Must be authorized by admin
-    fn updt_price(e: Env, n: u32, d: u32);
-
-    // Get the current price
-    fn get_price(e: Env) -> Price;
-
-    fn get_sell(e: Env) -> BytesN<32>;
-
-    fn get_buy(e: Env) -> BytesN<32>;
-}
-
-pub struct SingleOffer;
-
 #[contractimpl]
-impl SingleOfferTrait for SingleOffer {
-    fn initialize(
+impl SingleOffer {
+    // Creates the offer for seller for the given token pair and initial price.
+    // See comment above the `Offer` struct for information on pricing.
+    pub fn create(
         e: Env,
-        admin: Identifier,
+        seller: Address,
         sell_token: BytesN<32>,
         buy_token: BytesN<32>,
-        n: u32,
-        d: u32,
+        sell_price: u32,
+        buy_price: u32,
     ) {
-        if has_administrator(&e) {
-            panic!("admin is already set");
+        if e.storage().has(&DataKey::Offer) {
+            panic!("offer is already created");
         }
-
-        if d == 0 {
-            panic!("d is zero but cannot be zero");
+        if buy_price == 0 || sell_price == 0 {
+            panic!("zero price is not allowed");
         }
-
-        write_administrator(&e, admin);
-
-        put_sell_token(&e, sell_token);
-        put_buy_token(&e, buy_token);
-        put_price(&e, Price { n, d });
+        // Authorize the `create` call by seller to verify their identity.
+        seller.require_auth();
+        write_offer(
+            &e,
+            &Offer {
+                seller,
+                sell_token,
+                buy_token,
+                sell_price,
+                buy_price,
+            },
+        );
     }
 
-    fn trade(e: Env, to: Identifier, min: i128) {
-        let balance_buy_token = get_balance_buy(&e);
-
-        let price = load_price(&e);
-
-        let amount = balance_buy_token * price.d as i128 / price.n as i128;
-
-        if amount < min {
-            panic!("will receive less than min");
-        }
-
-        transfer_sell(&e, to, amount);
-
-        let admin = read_administrator(&e);
-        transfer_buy(&e, admin, balance_buy_token);
+    // Trades `buy_token_amount` of buy_token from buyer for `sell_token` amount
+    // defined by the price.
+    // Buyer needs to authorize the `trade` call and internal `xfer` call to
+    // the contract address.
+    pub fn trade(e: Env, buyer: Address, buy_token_amount: i128) {
```

Comment:
I see you removed the min logic. Was that intentional?

Reply:
I thought it's unnecessary, but now I realize that it's possible to frontrun a trade with a price update, which is bad. I'll commit a fix soon.